# Lab 11/ Spotify data set 🤢🤢🤢

Minakov Nikita, csse 1707DA

@minakov.nv

In [1]:
from __future__ import print_function, division

%matplotlib inline

import numpy as np

import pandas as pd
import nsfg
import first
import thinkstats2
import thinkplot

import matplotlib.pyplot as plt
from collections import Counter
import seaborn as sns

from scipy import stats
import statistics 
import brfss
import random

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

import statsmodels.formula.api as smf

## Multiple regression


In [2]:
db_read = pd.read_csv('SpotifyFeatures.csv', delimiter = ',')
db_read

genre               artist_name  \
0       Movie            Henri Salvador   
1       Movie         Martin & les fées   
2       Movie           Joseph Williams   
3       Movie            Henri Salvador   
4       Movie              Fabien Nataf   
5       Movie            Henri Salvador   
6       Movie         Martin & les fées   
7       Movie               Laura Mayne   
8       Movie                    Chorus   
9       Movie       Le Club des Juniors   
10      Movie         Leopold Stokowski   
11      Movie              Randy Newman   
12      Movie      Idoles De La Musique   
13      Movie                    Chorus   
14      Movie        Richard M. Sherman   
15      Movie               Michel Roux   
16      Movie        Jean Claude Corbel   
17      Movie             Bernard Minet   
18      Movie            Henri Salvador   
19      Movie                    Amalya   
20      Movie         Martin & les fées   
21      Movie                  Dorothée   
22      Movie            Henri Salvador   
23      Movie        Jean Claude Corbel   
24      Movie            Henri Salvador   
25      Movie                 Jean Piat   
26      Movie            Henri Salvador   
27      Movie              Lucie Dolene   
28      Movie            Henri Salvador   
29      Movie                    Chorus   
...       ...                       ...   
232695   Soul             Amy Winehouse   
232696   Soul                  Autograf   
232697   Soul                     BAYNK   
232698   Soul              Donell Jones   
232699   Soul        The Isley Brothers   
232700   Soul              Willie Hutch   
232701   Soul                       Kem   
232702   Soul                 B.B. King   
232703   Soul                  Ella Vos   
232704   Soul                Joss Stone   
232705   Soul               Irma Thomas   
232706   Soul               Iris Temple   
232707   Soul             Tamar Braxton   
232708   Soul             Rexx Life Raj   
232709   Soul                Joss Stone   
232710   Soul                       SiR   
232711   Soul                The Knocks   
232712   Soul        Great Good Fine Ok   
232713   Soul               Keith Sweat   
232714   Soul                    H-Town   
232715   Soul                Emily King   
232716   Soul               John Legend   
232717   Soul                     Belly   
232718   Soul              Muddy Waters   
232719   Soul        Bobby "Blue" Bland   
232720   Soul                     Slave   
232721   Soul  Jr Thomas & The Volcanos   
232722   Soul              Muddy Waters   
232723   Soul                   R.LUM.R   
232724   Soul            Mint Condition   

                                               track_name  \
0                             C'est beau de faire un Show   
1                        Perdu d'avance (par Gad Elmaleh)   
2                          Don't Let Me Be Lonely Tonight   
3                          Dis-moi Monsieur Gordon Cooper   
4                                               Ouverture   
5                          Le petit souper aux chandelles   
6       Premières recherches (par Paul Ventimila, Lori...   
7                                           Let Me Let Go   
8                                                   Helka   
9                               Les bisous des bisounours   
10      Symphony No.4 In E Minor Op.98 : IV. Allegro E...   
11      The Hanging (Maverick - Original Motion Pictur...   
12                                 J'ai demandé à la lune   
13                                          Mangala Aarti   
14                                           Keys of Love   
15           Les aventures d'Arsène Lupin: 813 - Partie 6   
16                                                  Diane   
17                                           Ultra Man 80   
18                                                Veunise   
19                                   Ladies And Gentlemen   
20      Tout est possible (par Anggun, Dany Brillant, ...   

In [3]:
db_read.corr

<bound method DataFrame.corr of         genre               artist_name  \
0       Movie            Henri Salvador   
1       Movie         Martin & les fées   
2       Movie           Joseph Williams   
3       Movie            Henri Salvador   
4       Movie              Fabien Nataf   
5       Movie            Henri Salvador   
6       Movie         Martin & les fées   
7       Movie               Laura Mayne   
8       Movie                    Chorus   
9       Movie       Le Club des Juniors   
10      Movie         Leopold Stokowski   
11      Movie              Randy Newman   
12      Movie      Idoles De La Musique   
13      Movie                    Chorus   
14      Movie        Richard M. Sherman   
15      Movie               Michel Roux   
16      Movie        Jean Claude Corbel   
17      Movie             Bernard Minet   
18      Movie            Henri Salvador   
19      Movie                    Amalya   
20      Movie         Martin & les fées   
21      Movie         

Here's popularity as a function of tempo

In [4]:
formula = 'popularity ~ tempo'
model = smf.ols(formula, data=db_read)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             popularity   R-squared:                       0.007
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     1538.
Date:                Thu, 14 Nov 2019   Prob (F-statistic):               0.00
Time:                        10:39:03   Log-Likelihood:            -1.0046e+06
No. Observations:              232725   AIC:                         2.009e+06
Df Residuals:                  232723   BIC:                         2.009e+06
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     35.5140      0.148    240.009      0.000      35.224      35.804
tempo          0.0477      0.001     39.223      0.000       0.045       0.050
==============================================================================
Omnibus:                     6340.918   Durbin-Watson:                   0.180
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5742.824
Skew:                          -0.333   Prob(JB):                         0.00
Kurtosis:                       2.614   Cond. No.                         479.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

We can extract the parameters.

In [5]:
inter = results.params['Intercept']
slope = results.params['tempo']
inter, slope

(35.51398509741364, 0.047706979213790385)

And the p-value of the slope estimate.

In [8]:
slope_pvalue = results.pvalues['tempo']
print(slope_pvalue)

0.0


And the coefficient of determination.

In [10]:
results.rsquared

0.00656730466239841

In [11]:
popM = db_read[db_read['genre'] == 'Pop']
movieM = db_read[db_read['genre'] == 'Movie']

The difference in popularity between Pop and movie songs.

In [12]:
diff_weight = popM.popularity.mean() - movieM.popularity.mean()
diff_weight

54.41657010219974

The difference in tempo between Pop and Movie songs.

In [13]:
diff_age = popM.tempo.mean() - movieM.tempo.mean()
diff_age

7.229074315522226

The age difference plausibly explains about half of the difference in popularity.

In [14]:
slope * diff_age

0.3448772981055648

Running a single regression with a categorical variable, `isGkey`:

In [15]:
db_read['isGkey'] = db_read.key == 'G#'
formula = 'popularity ~ isGkey'
results = smf.ols(formula, data=db_read).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             popularity   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     152.5
Date:                Thu, 14 Nov 2019   Prob (F-statistic):           5.05e-35
Time:                        10:44:28   Log-Likelihood:            -1.0053e+06
No. Observations:              232725   AIC:                         2.011e+06
Df Residuals:                  232723   BIC:                         2.011e+06
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         41.0046      0.039   1051.812      0.000      40.928      41.081
isGkey[T.True]     1.8863      0.153     12.349      0.000       1.587       2.186
==============================================================================
Omnibus:                     6638.002   Durbin-Watson:                   0.167
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5948.404
Skew:                          -0.337   Prob(JB):                         0.00
Kurtosis:                       2.600   Cond. No.                         4.07
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Now finally running a multiple regression:

In [16]:
formula = 'popularity ~ isGkey + tempo'
results = smf.ols(formula, data=db_read).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             popularity   R-squared:                       0.007
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     845.5
Date:                Thu, 14 Nov 2019   Prob (F-statistic):               0.00
Time:                        10:44:48   Log-Likelihood:            -1.0045e+06
No. Observations:              232725   AIC:                         2.009e+06
Df Residuals:                  232722   BIC:                         2.009e+06
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         35.3956      0.148    238.783      0.000      35.105      35.686
isGkey[T.True]     1.8738      0.152     12.307      0.000       1.575       2.172
tempo              0.0477      0.001     39.210      0.000       0.045       0.050
==============================================================================
Omnibus:                     6324.801   Durbin-Watson:                   0.182
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5738.581
Skew:                          -0.333   Prob(JB):                         0.00
Kurtosis:                       2.615   Cond. No.                         503.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

As expected, when we control for popularity, the apparent difference due to `isGkey` is cut in half.

If we add age squared, we can control for a quadratic relationship between popularity and tempo.

In [17]:
db_read['isGkey'] = db_read.energy**2
formula = 'popularity ~ isGkey + tempo + tempo'
results = smf.ols(formula, data=db_read).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             popularity   R-squared:                       0.036
Model:                            OLS   Adj. R-squared:                  0.036
Method:                 Least Squares   F-statistic:                     4363.
Date:                Thu, 14 Nov 2019   Prob (F-statistic):               0.00
Time:                        10:45:55   Log-Likelihood:            -1.0010e+06
No. Observations:              232725   AIC:                         2.002e+06
Df Residuals:                  232722   BIC:                         2.002e+06
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     33.6708      0.147    228.473      0.000      33.382      33.960
isGkey        11.3655      0.134     84.505      0.000      11.102      11.629
tempo          0.0252      0.001     20.515      0.000       0.023       0.028
==============================================================================
Omnibus:                     6838.613   Durbin-Watson:                   0.218
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5271.880
Skew:                          -0.282   Prob(JB):                         0.00
Kurtosis:                       2.526   Cond. No.                         502.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Data Mining

And we can search for variables with explanatory power.

Because we don't clean most of the variables, we are probably missing some good ones.

In [18]:
import patsy

def GoMining(df):
    """Searches for variables that predict birth weight.

    df: DataFrame of pregnancy records

    returns: list of (rsquared, variable name) pairs
    """
    variables = []
    for name in df.columns:
        try:
            if df[name].var() < 1e-7:
                continue

            formula = 'totalwgt_lb ~ agepreg + ' + name
            formula = formula.encode('ascii')

            model = smf.ols(formula, data=df)
            if model.nobs < len(df)/2:
                continue

            results = model.fit()
        except (ValueError, TypeError, patsy.PatsyError):
            continue
        
        variables.append((results.rsquared, name))

    return variables

In [22]:
variables = GoMining(db_read)

The following functions report the variables with the highest values of $R^2$.

In [19]:
import re

def ReadVariables():
    """Reads Stata dictionary files for NSFG data.

    returns: DataFrame that maps variables names to descriptions
    """
    vars1 = thinkstats2.ReadStataDct('2002FemPreg.dct').variables
    vars2 = thinkstats2.ReadStataDct('2002FemResp.dct').variables

    all_vars = vars1.append(vars2)
    all_vars.index = all_vars.name
    return all_vars

def MiningReport(variables, n=30):
    """Prints variables with the highest R^2.

    t: list of (R^2, variable name) pairs
    n: number of pairs to print
    """
    all_vars = ReadVariables()

    variables.sort(reverse=True)
    for r2, name in variables[:n]:
        key = re.sub('_r$', '', name)
        try:
            desc = all_vars.loc[key].desc
            if isinstance(desc, pd.Series):
                desc = desc[0]
            print(name, r2, desc)
        except (KeyError, IndexError):
            print(name, r2)

Some of the variables that do well are not useful for prediction because they are not known ahead of time.

In [24]:
MiningReport(variables)

Combining the variables that seem to have the most explanatory power.

In [27]:
formula = ('popularity ~ tempo  + key=="G"')
results = smf.ols(formula, data=db_read).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             popularity   R-squared:                       0.007
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     836.0
Date:                Thu, 14 Nov 2019   Prob (F-statistic):               0.00
Time:                        10:50:04   Log-Likelihood:            -1.0045e+06
No. Observations:              232725   AIC:                         2.009e+06
Df Residuals:                  232722   BIC:                         2.009e+06
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept             35.6667      0.149    240.146      0.000      35.376      35.958
key == "G"[T.True]    -1.3643      0.118    -11.514      0.000      -1.597      -1.132
tempo                  0.0477      0.001     39.248      0.000       0.045       0.050
==============================================================================
Omnibus:                     6340.008   Durbin-Watson:                   0.181
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5748.423
Skew:                          -0.333   Prob(JB):                         0.00
Kurtosis:                       2.614   Cond. No.                         484.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Logistic regression

In [30]:
#import first
#live, firsts, others = first.MakeFrames()
#live = live[live.prglngth>30]
#live['boy'] = (live.babysex==1).astype(int)

db_read['pop50'] = (db_read.popularity == 50).astype(int)

In [31]:
model = smf.logit('pop50 ~ tempo', data=db_read)
results = model.fit()
results.summary()

Optimization terminated successfully.
         Current function value: 0.110459
         Iterations 8


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  pop50   No. Observations:               232725
Model:                          Logit   Df Residuals:                   232723
Method:                           MLE   Df Model:                            1
Date:                Thu, 14 Nov 2019   Pseudo R-squ.:               0.0003538
Time:                        11:41:16   Log-Likelihood:                -25707.
converged:                       True   LL-Null:                       -25716.
Covariance Type:            nonrobust   LLR p-value:                 1.992e-05
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -3.9605      0.054    -72.881      0.000      -4.067      -3.854
tempo          0.0019      0.000      4.280      0.000       0.001       0.003
==============================================================================
"""

Here are the variables that seemed most promising.

In [33]:
formula = 'pop50 ~ tempo + valence'
model = smf.logit(formula, data=db_read)
results = model.fit()
results.summary()

Optimization terminated successfully.
         Current function value: 0.110451
         Iterations 8


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  pop50   No. Observations:               232725
Model:                          Logit   Df Residuals:                   232722
Method:                           MLE   Df Model:                            2
Date:                Thu, 14 Nov 2019   Pseudo R-squ.:               0.0004271
Time:                        11:43:27   Log-Likelihood:                -25705.
converged:                       True   LL-Null:                       -25716.
Covariance Type:            nonrobust   LLR p-value:                 1.697e-05
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -3.9947      0.057    -69.810      0.000      -4.107      -3.883
tempo          0.0018      0.000      3.985      0.000       0.001       0.003
valence        0.1036      0.053      1.943      0.052      -0.001       0.208
==============================================================================
"""

In [34]:
endog = pd.DataFrame(model.endog, columns=[model.endog_names])
exog = pd.DataFrame(model.exog, columns=model.exog_names)

The baseline prediction strategy is to guess "pop50".

In [35]:
actual = endog['pop50']
baseline = actual.mean()
baseline

0.023267805349661618

If we use the previous model, we can compute the number of predictions we get right.

In [36]:
predict = (results.predict() >= 0.5)
true_pos = predict * actual
true_neg = (1 - predict) * (1 - actual)
sum(true_pos), sum(true_neg)

(0.0, 227310.0)

Accuracy:

In [37]:
acc = (sum(true_pos) + sum(true_neg)) / len(actual)
acc

0.9767321946503383

To make a prediction for an individual, we have to get their information into a `DataFrame`.

In [42]:
columns = ['popularity', 'valence', 'tempo', 'liveness']
new = pd.DataFrame([[35, 39, 3, 10]], columns=columns)
y = results.predict(new)
y

0    0.512404
dtype: float64

This person has a 51% chance of having a popularity 50 (according to the model).